clients/src/main/java/org/apache/kafka/clients/NetworkClientUtils.java

```diff
@@ -60,17 +60,16 @@ public static boolean awaitReady(KafkaClient client, Node node, Time time, long
             throw new IllegalArgumentException("Timeout needs to be greater than 0");
         }
         long startTime = time.milliseconds();
-        long expiryTime = startTime + timeoutMs;
 
         if (isReady(client, node, startTime) ||  client.ready(node, startTime))
             return true;
 
         long attemptStartTime = time.milliseconds();
-        while (!client.isReady(node, attemptStartTime) && attemptStartTime < expiryTime) {
+        while (!client.isReady(node, attemptStartTime) && attemptStartTime - startTime < timeoutMs) {
             if (client.connectionFailed(node)) {
                 throw new IOException("Connection to " + node + " failed.");
             }
-            long pollTimeout = expiryTime - attemptStartTime;
+            long pollTimeout = (startTime - attemptStartTime) + timeoutMs;
```

Comment:
This would still overflow if `timeoutMs` is `MAX_VALUE`, right?

Reply:
The `startTime` is set once at the beginning of the method while the `attemptStartTime` is initialized just before the first attempt and then updated again after every iteration. So the `attemptStartTime` is always greater than the `startTime` and therefore the quantity being added to the `timeoutMs` here is actually negative. 

But I see how that's confusing, I'll refactor the expression to make this more clear

Reply:
Ah, thanks!